### DISCUSSION TITLE
Unable to sync due to 'HTTP request returned status code 429' - The request has been throttled

### DISCUSSION THREAD

**Discussion by dtto** [abraunegg/onedrive#1837] [Q&A] [answered]
Hello and thank you for all your work on this software. I am unsure whether this is a bug or simply a misconfiguration on my end, so I'm posting here in discussions. If I need to move this to issues, happy to do so.

I have installed onedrive v2.4.15-1-np1 on a debian bullseye host. 

```
# cat .config/onedrive/config
sync_business_shared_folders = "true"
enable_logging = "true"
log_dir = "/var/log/onedrive/"
sync_dir = "/data/onedrive/"
check_nomount = "true"
check_nosync = "true"
```

I attempt to run the command `onedrive --synchronize --verbose` And the output is:

```
Using 'user' Config Dir: /home/UID/.config/onedrive
Using 'system' Config Dir: /etc/onedrive
Configuration file successfully loaded
Using logfile dir: /var/log/onedrive/
Initializing the OneDrive API ...
Configuring Global Azure AD Endpoints
Opening the item database ...
All operations will be performed in: /data/onedrive/
Application version: v2.4.15-1~np1
Account Type: business
Default Drive ID: b!KbBLqycoD02sU9GR660ES1Pl5nbtc6hNm7nTsmB_7a46DwdEgDXZSKCeemYLHNKy
Default Root ID: 01KH4S2WF6Y2GOVW7725BZO354PWSELRRZ
Remaining Free Space: 1099480515269
Fetching details for OneDrive Root
OneDrive Root exists in the database
Initializing the Synchronization Engine ...
Syncing changes from OneDrive ...
Applying changes of Path ID: 01KH4S2WF6Y2GOVW7725BZO354PWSELRRZ
Updated Remaining Free Space: 1099480515269
Number of items from OneDrive to process: 0
Attempting to sync OneDrive Business Shared Folders
Syncing this OneDrive Business Shared Folder: SHARENAME
OneDrive Business Shared Folder - Shared By:  XXXX
Applying changes of Path ID: 01XELLNHFBIGWPVRK55ZB23GHO5XG2YZPB
Adding OneDrive root details for processing
Adding OneDrive folder details for processing
Adding 6 OneDrive items for processing from OneDrive folder
```

This continues for some time, but then I run into three 429 errors, as follows:

```
Thread sleeping due to 'HTTP request returned status code 429' - The request has been throttled
Sleeping for 42 seconds
Adding 3 OneDrive items for processing from XXX
Thread sleeping due to 'HTTP request returned status code 429' - The request has been throttled
Sleeping for 205 seconds
Adding 1 OneDrive items for processing from xxxx
... continues ...
Thread sleeping due to 'HTTP request returned status code 429' - The request has been throttled
Sleeping for 33 seconds

ERROR: Microsoft OneDrive API returned an error with the following message:
  Error Message:    HTTP request returned status code 429 ()
  Error Reason:     The request has been throttled
  Error Timestamp:  2022-02-07T10:19:33
  API Request ID:   1311d55b-c5dc-4821-8782-adb57cfb0145
  Calling Function: queryForChildren()

Giving up on sync after three attempts: JSONValue is not an object
std.json.JSONException@std/json.d(308): JSONValue is not an object
----------------
??:? pure @safe bool std.exception.enforce!(std.json.JSONException).enforce!(bool).enforce(bool, lazy const(char)[], immutable(char)[], ulong) [0x55c2a31b97a2]
??:? inout pure ref @safe inout(std.json.JSONValue) std.json.JSONValue.opIndex(immutable(char)[]) [0x7fc91ad65220]
??:? [0x55c2a31ec981]
??:? [0x55c2a31ecd4d]
??:? [0x55c2a31ecd4d]
??:? [0x55c2a31ecd4d]
??:? [0x55c2a31ecd4d]
??:? [0x55c2a31d47bc]
??:? [0x55c2a31c3a83]
??:? [0x55c2a31c22cf]
??:? [0x55c2a320de70]
??:? [0x55c2a3205604]
??:? void rt.dmain2._d_run_main2(char[][], ulong, extern (C) int function(char[][])*).runAll() [0x7fc91aa5907b]
??:? _d_run_main2 [0x7fc91aa58e97]
??:? _d_run_main [0x7fc91aa58ced]
??:? __libc_start_main [0x7fc91a661d09]
??:? [0x55c2a319a029]
std.net.curl.CurlException@std/net/curl.d(4402): A libcurl function was given a bad argument on handle null
----------------
??:? pure @safe bool std.exception.enforce!(std.net.curl.CurlException).enforce!(bool).enforce(bool, lazy const(char)[], immutable(char)[], ulong) [0x7fc91ad0bc7c]
??:? void std.net.curl.HTTP.clearRequestHeaders() [0x7fc91ad72d6f]
??:? [0x55c2a32071c3]
??:? void rt.dmain2._d_run_main2(char[][], ulong, extern (C) int function(char[][])*).runAll() [0x7fc91aa5907b]
??:? _d_run_main2 [0x7fc91aa58e97]
??:? _d_run_main [0x7fc91aa58ced]
??:? __libc_start_main [0x7fc91a661d09]
??:? [0x55c2a319a029]
```

This sync is running in the middle of the night at 3AM with no other syncs running from any user.

Any advice on how to resolve this would be greatly appreciated. And thank you again for all the work.

**Comment by abraunegg** [comment, marked as answer]
@dtto 
> I have installed onedrive v2.4.15-1-np1 on a debian bullseye host.
> ...
> `Giving up on sync after three attempts: JSONValue is not an object`

This is because you have installed using the Debian package from the OpenSuSE Build Service.

Because you are using the package, you are impacted by a SharePoint bug fixed post v2.4.15 release.

To resolve your issue:
1. Uninstall the 'onedrive' package
2. Build and install from source

Once v2.4.16 is released, you can go back to using the OpenSuSE Build Service packages.

**Reply by dtto** [reply]
> Once v2.4.16 is released, you can go back to using the OpenSuSE Build Service packages.

Thank you! I think I saw something in another post mentioning this and I had hoped this was the solution. Do you know the approximate timeframe for the .16 release?

Again, really appreciate your help with this and all your work on this.

**Reply by abraunegg** [reply]
> Thank you! I think I saw something in another post mentioning this and I had hoped this was the solution. Do you know the approximate timeframe for the .16 release?

Will be a few months - quite a bit of work & features to include

**Comment by dtto** [comment]
> To resolve your issue:
>
> Uninstall the 'onedrive' package
> Build and install from source

I am now running `onedrive v2.4.15-16-g3c119d3`, but I am still not able to get a reliable sync for two business share folders. After the upgrade, I removed the contents of the local onedrive folder and ran:

`onedrive --synchronize --download-only`

The sync will start, but will not finish. First a bunch of files are synced, with Filename ...done, but then:

` Downloading file PATH/TO/FILE ... ERROR: Query of OneDrive for file details failed.`

Then a bunch of similar failures, followed by the dreaded

`Thread sleeping due to 'HTTP request returned status code 429' - The request has been throttled`

And then more of the of the previous errors.

At the end of the sync, my local OneDrive folder is still missing a bunch of files. 

Do I just need to run `onedrive --synchronize` repeatedly until I get a complete copy? I thought maybe it was because the remote is so large, so I removed the largest folder using a skip_dir config entry, followed by a `--resync`, but even without that directory it's still not finishing.

**Comment by abraunegg** [comment]
@dtto 
Suggest that you do the following:

1. Make sure that there are no other 'onedrive' binaries lurking on your system - search and find them all, then ensure that only the version built from master exists. You may have older binaries still lurking around. Make sure you also uninstall any 'package' version.
2. Rather than continuing this discussion, open an issue ticket and provide all the required details including a verbose debug log. The verbose debug log is required so that if this is an application issue, the issue can be quickly identified and fixed.
